targets/psa_switch/externs/psa_checksum.cpp

```diff
@@ -0,0 +1,181 @@
+/* Copyright 2021 SYRMIA LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/*
+ * Dusan Krdzic (dusan.krdzic@syrmia.com)
+ *
+ */
+
+
+#include "psa_checksum.h"
+#include <bm/bm_sim/calculations.h>
+
+namespace {
+
+static const unsigned char hex_digits[] = {'0','1','2','3','4','5','6','7','8','9','A','B','C','D','E','F'};
+
+uint8_t convertHexStrToU8(std::string hex) {
+  int a;
+  std::istringstream(hex) >> std::hex >> a;
+  return static_cast<uint8_t>(a);
+}
+
+std::string prepareDataForHash(bm::Data &input) {
+  std::string hex_ascii;
+  std::string hex_hex;
+  const uint8_t base = 16;
+
+  while (input != 0) {
+    hex_ascii += hex_digits[input % base];
+    input.divide(input, base);
+  }
+  std::reverse(hex_ascii.begin(), hex_ascii.end());
+
+  if(hex_ascii.size() % 2 != 0) {
+    hex_ascii.insert(hex_ascii.size()-1, 1, '0');
+  }
```

Comment:
For odd number of digits in hex string insert '0'. For example "1C3" -> "1C03". This is important because hash algorithm takes 1B (2 hex digits). "1C3" will be processed as 0x1C 0x03 in hash algorithm.